File: Leetcode/Python3/1345. Jump Game IV/solution.py
# https://leetcode.com/problems/jump-game-iv/


from collections import defaultdict, deque


class Solution:
    def minJumps(self, arr: List[int]) -> int:
        index_dict = defaultdict(list)
        dp = [float("inf") for _ in range(len(arr))]
        for idx, num in enumerate(arr):
            index_dict[num].append(idx)

        queue = deque([0])
        dp[0] = 0   # Start point
        while queue:
            cur = queue.popleft()
            if cur == len(arr) - 1:  # We use BFS so first reach is fastest jump
                break
            # Use backtracking
            if cur - 1 >= 0 and dp[cur] + 1 < dp[cur - 1]:
                queue.append(cur - 1)
                dp[cur - 1] = dp[cur] + 1
            if cur + 1 < len(arr) and dp[cur] + 1 < dp[cur + 1]:
                queue.append(cur + 1)
                dp[cur + 1] = dp[cur] + 1
            for idx in index_dict[arr[cur]]:
                if idx != cur and dp[cur] + 1 < dp[idx]:
                    queue.append(idx)
                    dp[idx] = dp[cur] + 1
            index_dict[arr[cur]] = []   # After jump same value index clear list for prevent revisit
        print(dp)
        return dp[-1]

File: Leetcode/C++/1345. Jump Game IV/solution.cpp
// https://leetcode.com/problems/jump-game-iv/description/

#include <queue>
#include <unordered_map>
#include <vector>

using namespace std;

class Solution
{
public:
    int minJumps(vector<int> &arr)
    {
        int len = arr.size();
        unordered_map<int, vector<int>> index_hash;
        for (int i = 0; i < len; i++)
        {
            index_hash[arr[i]].push_back(i);
        }

        vector<int> dp(len, -1); // Initialize vector -1
        dp[0] = 0;
        queue<int> q;
        q.push(0); // Strat point
        while (!(q.empty()))
        {
            for (int i = q.size(); 0 < i; i--)
            {
                int cur = q.front();
                if (cur == len - 1) // Reach
                {
                    return dp[len - 1];
                }
                vector<int> &next = index_hash[arr[cur]];
                // Combine into next vector and process at the same time
                next.push_back(cur - 1);
                next.push_back(cur + 1);
                for (int j = 0; j < next.size(); j++)
                {
                    if (next[j] >= 0 && next[j] < len && dp[next[j]] == -1)
                    {
                        q.push(next[j]);
                        dp[next[j]] = dp[cur] + 1;
                    }
                }
                next = {}; // For prevent revisit clear vector
                q.pop();
            }
        }
        return 0;
    }
};